service/controller/resource/subnet/resource.go

```diff
@@ -0,0 +1,277 @@
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			return nil
+		}
+	}
+
+	return nil
+}
+
+// This functions decides whether or not the ARM deployment is out of date.
+// For that, we use the Generation field from the AzureCluster CR. This Generation field should change when there is a change in the CR.
+func (r *Resource) isDeploymentOutOfDate(ctx context.Context, cr capzv1alpha3.AzureCluster, currentDeployment azureresource.DeploymentExtended) (bool, error) {
+	crVersion := strconv.FormatInt(cr.ObjectMeta.Generation, 10)
+	currentParams, ok := currentDeployment.Properties.Parameters.(map[string]interface{})
+	if !ok {
+		return false, microerror.Maskf(wrongTypeError, "expected '%T', got '%T'", map[string]interface{}{}, currentDeployment.Properties.Parameters)
+	}
+
+	deploymentVersion, ok := currentParams["azureClusterVersion"].(map[string]interface{})["value"].(string)
+	if !ok {
+		return false, microerror.Maskf(wrongTypeError, "expected 'string', got '%T'", currentDeployment.Properties.Parameters)
+	}
+
+	r.logger.LogCtx(ctx, "message", "Checking if deployment is out of date", "azureClusterVersion", crVersion, "deploymentParameter", currentParams["azureClusterVersion"])
+
+	return crVersion != deploymentVersion, nil
+}
+
+func (r *Resource) getDeploymentParameters(ctx context.Context, clusterID, azureClusterVersion, virtualNetworkName string, allocatedSubnet *capzv1alpha3.SubnetSpec) (map[string]interface{}, error) {
+	return map[string]interface{}{
+		"azureClusterVersion": azureClusterVersion,
+		"nateGatewayName":     "workers-nat-gw",
+		"nodepoolName":        allocatedSubnet.Name,
+		"routeTableName":      fmt.Sprintf("%s-%s", clusterID, "RouteTable"),
+		"securityGroupName":   fmt.Sprintf("%s-%s", clusterID, "WorkerSecurityGroup"),
+		"subnetCidr":          allocatedSubnet.CidrBlock,
+		"virtualNetworkName":  virtualNetworkName,
+	}, nil
+}
+
+// EnsureDeleted is a noop since the deletion of deployments is redirected to
+// the deletion of resource groups because they garbage collect them.
+func (r *Resource) createDeployment(ctx context.Context, deploymentsClient *azureresource.DeploymentsClient, resourceGroup, deploymentName string, desiredDeployment azureresource.Deployment) error {
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring subnets deployments")
+
+	res, err := deploymentsClient.CreateOrUpdate(ctx, resourceGroup, deploymentName, desiredDeployment)
+	if err != nil {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment failed; deployment: %#v", desiredDeployment), "stack", microerror.JSON(microerror.Mask(err)))
+		return microerror.Mask(err)
+	}
+	deploymentExtended, err := deploymentsClient.CreateOrUpdateResponder(res.Response())
+	if err != nil {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment failed; deployment: %#v", deploymentExtended), "stack", microerror.JSON(microerror.Mask(err)))
```

Comment:
Same for reusing error variable in logging and return statement (not fully sure what's better).